### DISCUSSION TITLE
subscribing for a side-effect?

### DISCUSSION THREAD

**Discussion by keturn** [kennyfrc/cami.js#11] [Q&A]
Hi, I've been experimenting with Cami, and it's going pretty well so far.

I started using `query` to fetch a resource, and I'm trying to separate some functionality that happens in response to that.

One component displays the resource as-is. That's working fine.

I want to be able to have a second object that, when it notices that `Foo` resource is available, processes it to come up with a new `Foobar`.

It feels like this should be basically 

```js
this.foo = this.query(…)
this.foo.onValue(result => processFooIntoFoobar(result.data))
```
but there's some nuance that I'm missing. That callback seems to only be getting the new value in its `pending` status, not when it is finished with the data.

**Comment by kennyfrc** [comment]
Hi @keturn! Glad to hear you're liking it so far.

The short answer on why `query` and `onValue` don't work together is that `query` returns an `ObservableState` (similar to signals in preact / solid) and `onValue` expects an `ObservableStream` (similar to RXJS streams / Bacon.js), so there are actually two paradigms within cami.

There are two possible ways to accomplish what you want to do:

- The "signals" way: `query` + `effect` (see here: [this.effect](https://camijs.com/api/reactive_element/#reactiveelementobservableattributesattributes-void))
- The "streams" way: `http` + `onValue` (see here: [http stream](https://camijs.com/api/http/))

The way you can distinguish the difference between the two is as follows:

- `effect` listens to changes in *a* state (singular, non time-based)
- `onValue` listens to many state emissions over time (plural, time-based)

Since `query` just modifies *a* state, `onValue` only got the original state emission, and was not aware of the state change that happened later. `http` is designed to emit *many* states over time.

Some jsfiddle examples:

- Subscribing to `query`'s return value outside of the `template` by using `effect`: [link](https://jsfiddle.net/kennyfrc12/cdzhtpLf/12/)
- Subscribing to `http` by using `onValue`: [link](https://jsfiddle.net/kennyfrc12/xLe5dj8w/10/)

Lastly, you can also do this pipeline if desired: `http` -> `onValue` -> `state`, similar to what Ryan Carniato described here: [link](https://ryansolid.medium.com/yes-ive-seen-flyd-1e83e6fa8b2) (in that comment, you can replace 'Flyd' with http / onValue and 'Solid State Object' with query / effect / state.

**Reply by kennyfrc** [reply]
Although fwiw, I usually do `query` then `store.dispatch` for those type of things. I don't use `effect` typically. 

I also rarely use `stream` and `http` -- it's most useful for real time event streams
